File: iosApp/iosApp/view/widgets/smartText/SmartTextSheets.swift
import SwiftUI
import shared

struct SmartTextSheetHost: View {

    @ObservedObject var store: SmartTextStore

    @Environment(\.colorScheme) private var scheme

    private var palette: SmartTextPalette { SmartTextPalette.of(scheme) }

    private var commands: SmartTextCommands { SmartTextCommands.shared }

    private var catalog: SmartTextCatalog { SmartTextCatalog.shared }

    var body: some View {
        NavigationStack {
            content
                .padding(.horizontal, 20)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .presentationDetents([.medium, .large])
        .presentationDragIndicator(.visible)
        .background(palette.surface)
    }

    @ViewBuilder
    private var content: some View {
        switch store.sheet {
        case .textStyle: textStyleSheet
        case .textColor: colorSheet(title: "Text colour", palette: catalog.textColors, isBackground: false)
        case .backgroundColor:
            colorSheet(title: "Background colour", palette: catalog.highlightColors, isBackground: true)
        case .fontSize: fontSizeSheet
        case .align: alignSheet
        case .table: tableSheet
        case .link: linkSheet
        case .findReplace: findReplaceSheet
        case .none: EmptyView()
        }
    }

    private func title(_ text: String) -> some View {
        Text(text)
            .font(.system(size: 17, weight: .semibold))
            .foregroundColor(palette.onSurface)
            .padding(.top, 16)
            .padding(.bottom, 10)
    }

    private var textStyleSheet: some View {
        VStack(alignment: .leading, spacing: 0) {
            title("Text style")
            ScrollView {
                VStack(alignment: .leading, spacing: 2) {
                    ForEach(catalog.paragraphStyles, id: \.self) { style in
                        let selected = style == store.toolbar.paragraphStyle
                        Button {
                            store.dispatch(commands.setParagraphStyle(style: style))
                            store.sheet = .none
                        } label: {
                            Text(catalog.paragraphLabel(style: style))
                                .font(
                                    .system(
                                        size: 15 * CGFloat(min(style.relativeSize, 1.6)),
                                        weight: SmartTextStyleMapper.swiftUIWeight(Int(style.weight))
                                    )
                                )
                                .foregroundColor(selected ? palette.accent : palette.onSurface)
                                .frame(maxWidth: .infinity, alignment: .leading)
                                .padding(.horizontal, 12)
                                .padding(.vertical, 10)
                                .background(
                                    RoundedRectangle(cornerRadius: 8)
                                        .fill(selected ? palette.accentSoft : .clear)
                                )
                        }
                        .buttonStyle(.plain)
                    }
                }
            }
        }
    }

    private func colorSheet(title text: String, palette colors: [String], isBackground: Bool) -> some View {
        ColorSheetBody(
            title: text,
            swatches: colors,
            palette: palette,
            onPick: { hex in
                store.dispatch(
                    isBackground
                        ? commands.setBackgroundColor(color: hex)
                        : commands.setTextColor(color: hex)
                )
                store.sheet = .none
            }
        )
    }

    private var fontSizeSheet: some View {
        FontSizeSheetBody(
            initial: store.toolbar.fontSize.map { CGFloat($0.floatValue) } ?? 16,
            palette: palette,
            onApply: { size in
                store.dispatch(commands.setFontSize(size: size.map { KotlinFloat(float: Float($0)) }))
                store.sheet = .none
            }
        )
    }

    private var alignSheet: some View {
        VStack(alignment: .leading, spacing: 0) {
            title("Alignment")
            ForEach(catalog.alignments, id: \.self) { align in
                let selected = align == store.toolbar.align
                Button {
                    store.dispatch(commands.setAlignment(align: align))
                    store.sheet = .none
                } label: {
                    Text(catalog.alignLabel(align: align))
                        .font(.system(size: 15))
                        .foregroundColor(selected ? palette.accent : palette.onSurface)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .padding(.horizontal, 12)
                        .padding(.vertical, 12)
                        .background(
                            RoundedRectangle(cornerRadius: 8)
                                .fill(selected ? palette.accentSoft : .clear)
                        )
                }
                .buttonStyle(.plain)
            }
        }
    }

    private var tableSheet: some View {
        let rows = Int(store.focusedTable?.data.rowCount ?? 0)
        let columns = Int(store.focusedTable?.data.columnCount ?? 0)

        return VStack(alignment: .leading, spacing: 0) {
            title("Table")
            HStack(spacing: 10) {
                ForEach(0..<catalog.tablePresetRows.count, id: \.self) { index in
                    let presetRows = Int(truncating: catalog.tablePresetRows[index])
                    let presetColumns = Int(truncating: catalog.tablePresetColumns[index])
                    Button("\(presetRows) × \(presetColumns)") {
                        store.dispatch(
                            commands.insertTable(
                                rows: Int32(presetRows),
                                columns: Int32(presetColumns)
                            )
                        )
                        store.sheet = .none
                    }
                    .foregroundColor(palette.accent)
                    .padding(.horizontal, 14)
                    .padding(.vertical, 8)
                    .background(RoundedRectangle(cornerRadius: 8).fill(palette.accentSoft))
                }
            }
            .padding(.bottom, 16)

            if store.focusedTable != nil {
                tableRow("Add row") { commands.tableAddRow(at: Int32(rows)) }
                tableRow("Add column") { commands.tableAddColumn(at: Int32(columns)) }
                tableRow("Delete last row") { commands.tableDeleteRow(at: Int32(rows - 1)) }
                tableRow("Delete last column") { commands.tableDeleteColumn(at: Int32(columns - 1)) }
                tableRow("Toggle header row") { commands.tableToggleHeaderRow() }
                tableRow("Toggle header column") { commands.tableToggleHeaderColumn() }
            }
        }
    }

    private func tableRow(_ label: String, command: @escaping () -> TableCommand) -> some View {
        Button {
            guard let table = store.focusedTable else { return }
            store.dispatch(commands.tableAction(blockId: table.id, command: command()))
        } label: {
            Text(label)
                .font(.system(size: 15))
                .foregroundColor(palette.onSurface)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.vertical, 12)
        }
        .buttonStyle(.plain)
    }

    private var linkSheet: some View {
        LinkSheetBody(
            initial: store.toolbar.link ?? "",
            hasExisting: store.toolbar.link != nil,
            palette: palette,
            onApply: { url in
                store.dispatch(commands.setLink(url: url))
                store.sheet = .none
            }
        )
    }

    private var findReplaceSheet: some View {
        VStack(alignment: .leading, spacing: 10) {
            title("Find and replace")
            TextField(
                "Find",
                text: Binding(
                    get: { store.state.search.query },
                    set: { store.dispatch(commands.setSearchQuery(query: $0, matchCase: false)) }
                )
            )
            .textFieldStyle(.roundedBorder)

            Text(
                store.state.search.matches.isEmpty
                    ? "No matches"
                    : "\(Int(store.state.search.currentIndex) + 1) of \(store.state.search.matches.count)"
            )
            .font(.system(size: 13))
            .foregroundColor(palette.onSurfaceMuted)

            TextField(
                "Replace with",
                text: Binding(
                    get: { store.state.search.replacement },
                    set: { store.dispatch(commands.setReplacement(replacement: $0)) }
                )
            )
            .textFieldStyle(.roundedBorder)

            HStack(spacing: 8) {
                Button("Previous") { store.dispatch(commands.findPrevious()) }
                Button("Next") { store.dispatch(commands.findNext()) }
                Button("Replace") { store.dispatch(commands.replaceCurrent()) }
                Button("All") { store.dispatch(commands.replaceAll()) }
            }
            .foregroundColor(palette.accent)
            .padding(.top, 6)
            Spacer()
        }
    }
}

private struct FontSizeSheetBody: View {

    @State var size: CGFloat
    let palette: SmartTextPalette
    let onApply: (CGFloat?) -> Void

    init(initial: CGFloat, palette: SmartTextPalette, onApply: @escaping (CGFloat?) -> Void) {
        _size = State(initialValue: initial)
        self.palette = palette
        self.onApply = onApply
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("Font size")
                .font(.system(size: 17, weight: .semibold))
                .foregroundColor(palette.onSurface)
                .padding(.top, 16)
            Text("\(Int(size)) pt")
                .font(.system(size: 14))
                .foregroundColor(palette.onSurfaceMuted)
            Slider(value: $size, in: 10...48, step: 1) { editing in
                if !editing { onApply(size) }
            }
            .tint(palette.accent)
            Button("Reset to default") { onApply(nil) }
                .foregroundColor(palette.accent)
            Spacer()
        }
    }
}

private struct LinkSheetBody: View {

    @State var url: String
    let hasExisting: Bool
    let palette: SmartTextPalette
    let onApply: (String?) -> Void

    init(
        initial: String,
        hasExisting: Bool,
        palette: SmartTextPalette,
        onApply: @escaping (String?) -> Void
    ) {
        _url = State(initialValue: initial)
        self.hasExisting = hasExisting
        self.palette = palette
        self.onApply = onApply
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(hasExisting ? "Edit link" : "Insert link")
                .font(.system(size: 17, weight: .semibold))
                .foregroundColor(palette.onSurface)
                .padding(.top, 16)
            TextField("https://", text: $url)
                .textFieldStyle(.roundedBorder)
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled()
                .keyboardType(.URL)
            HStack(spacing: 12) {
                Button("Apply") { onApply(url.isEmpty ? nil : url) }
                    .foregroundColor(palette.onAccent)
                    .padding(.horizontal, 18)
                    .padding(.vertical, 10)
                    .background(RoundedRectangle(cornerRadius: 8).fill(palette.accent))
                if hasExisting {
                    Button("Remove link") { onApply(nil) }
                        .foregroundColor(palette.accent)
                }
            }
            Spacer()
        }
    }
}


/// Quick swatches plus the ColorSelector already used by Create Tag.
private struct ColorSheetBody: View {

    let title: String
    let swatches: [String]
    let palette: SmartTextPalette
    let onPick: (String?) -> Void

    @State private var picked: Color = .orange

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(title)
                .font(.system(size: 17, weight: .semibold))
                .foregroundColor(palette.onSurface)
                .padding(.top, 16)

            LazyVGrid(columns: Array(repeating: GridItem(.flexible()), count: 5), spacing: 12) {
                ForEach(swatches, id: \.self) { hex in
                    Button { onPick(hex) } label: {
                        Circle()
                            .fill(Color(smartTextHex: hex) ?? .gray)
                            .frame(width: 42, height: 42)
                            .overlay(Circle().stroke(palette.divider, lineWidth: 1))
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(hex)
                }
            }

            ScrollView {
                ColorSelector(selectedColor: $picked)
            }

            HStack(spacing: 12) {
                Button("Apply") { onPick(picked.tohexColor()) }
                    .foregroundColor(palette.onAccent)
                    .padding(.horizontal, 18)
                    .padding(.vertical, 10)
                    .background(RoundedRectangle(cornerRadius: 8).fill(palette.accent))
                Button("Remove colour") { onPick(nil) }
                    .foregroundColor(palette.accent)
            }
            .padding(.bottom, 16)
        }
    }
}

File: androidApp/src/main/java/com/app/pustakam/android/widgets/smartText/SmartTextSheets.kt
package com.app.pustakam.android.widgets.smartText

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.app.pustakam.android.widgets.colorPalete.ColorSelector
import com.app.pustakam.core.richtext.model.ParagraphStyle
import com.app.pustakam.core.richtext.model.TextAlign
import com.app.pustakam.core.richtext.presentation.SmartTextCatalog
import com.app.pustakam.core.richtext.presentation.SmartTextCommands
import com.app.pustakam.core.richtext.presentation.TableCommand

// which auxiliary sheet the editor is showing, if any
enum class SmartTextSheet {
    NONE,
    TEXT_STYLE,
    TEXT_COLOR,
    BACKGROUND_COLOR,
    FONT_SIZE,
    ALIGN,
    TABLE,
    LINK,
    FIND_REPLACE;

    companion object {
        fun fromIndex(index: Int): SmartTextSheet = when (index) {
            SmartTextCommands.SHEET_TEXT_STYLE -> TEXT_STYLE
            SmartTextCommands.SHEET_TEXT_COLOR -> TEXT_COLOR
            SmartTextCommands.SHEET_BACKGROUND_COLOR -> BACKGROUND_COLOR
            SmartTextCommands.SHEET_FONT_SIZE -> FONT_SIZE
            SmartTextCommands.SHEET_ALIGN -> ALIGN
            SmartTextCommands.SHEET_TABLE -> TABLE
            SmartTextCommands.SHEET_LINK -> LINK
            SmartTextCommands.SHEET_FIND_REPLACE -> FIND_REPLACE
            else -> NONE
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SmartTextSheetHost(
    sheet: SmartTextSheet,
    currentStyle: ParagraphStyle,
    currentAlign: TextAlign,
    currentFontSize: Float?,
    currentLink: String?,
    searchQuery: String,
    replacement: String,
    matchCount: Int,
    currentMatch: Int,
    tableRowCount: Int,
    tableColumnCount: Int,
    onDismiss: () -> Unit,
    onStyle: (ParagraphStyle) -> Unit,
    onAlign: (TextAlign) -> Unit,
    onColor: (String?) -> Unit,
    onFontSize: (Float?) -> Unit,
    onLink: (String?) -> Unit,
    onTable: (TableCommand) -> Unit,
    onInsertTable: (Int, Int) -> Unit,
    onSearchQuery: (String) -> Unit,
    onReplacement: (String) -> Unit,
    onFindNext: () -> Unit,
    onFindPrevious: () -> Unit,
    onReplaceCurrent: () -> Unit,
    onReplaceAll: () -> Unit
) {
    if (sheet == SmartTextSheet.NONE) return
    val colors = SmartTextTokens.colors
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = colors.surface,
        contentColor = colors.onSurface
    ) {
        Column(modifier = Modifier.padding(horizontal = 20.dp, vertical = 8.dp)) {
            when (sheet) {
                SmartTextSheet.TEXT_STYLE -> TextStyleSheet(currentStyle, onStyle)
                SmartTextSheet.TEXT_COLOR ->
                    ColorSheet("Text colour", SmartTextCatalog.textColors, onColor)

                SmartTextSheet.BACKGROUND_COLOR ->
                    ColorSheet("Background colour", SmartTextCatalog.highlightColors, onColor)

                SmartTextSheet.FONT_SIZE -> FontSizeSheet(currentFontSize, onFontSize)
                SmartTextSheet.ALIGN -> AlignSheet(currentAlign, onAlign)
                SmartTextSheet.TABLE ->
                    TableSheet(tableRowCount, tableColumnCount, onInsertTable, onTable)
                SmartTextSheet.LINK -> LinkSheet(currentLink, onLink)
                SmartTextSheet.FIND_REPLACE -> FindReplaceSheet(
                    query = searchQuery,
                    replacement = replacement,
                    matchCount = matchCount,
                    currentMatch = currentMatch,
                    onSearchQuery = onSearchQuery,
                    onReplacement = onReplacement,
                    onFindNext = onFindNext,
                    onFindPrevious = onFindPrevious,
                    onReplaceCurrent = onReplaceCurrent,
                    onReplaceAll = onReplaceAll
                )

                SmartTextSheet.NONE -> Unit
            }
        }
    }
}

@Composable
private fun SheetTitle(text: String) {
    Text(
        text = text,
        style = TextStyle(
            color = SmartTextTokens.colors.onSurface,
            fontSize = 17.sp,
            fontWeight = FontWeight.SemiBold
        ),
        modifier = Modifier.padding(bottom = 12.dp)
    )
}

@Composable
private fun TextStyleSheet(current: ParagraphStyle, onStyle: (ParagraphStyle) -> Unit) {
    val colors = SmartTextTokens.colors
    SheetTitle("Text style")
    Column(modifier = Modifier.padding(bottom = 20.dp)) {
        SmartTextCatalog.paragraphStyles.forEach { style ->
            val selected = style == current
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 2.dp)
                    .background(
                        if (selected) colors.accentSoft else Color.Transparent,
                        RoundedCornerShape(8.dp)
                    )
                    .clickable { onStyle(style) }
                    .padding(horizontal = 12.dp, vertical = 10.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = SmartTextCatalog.paragraphLabel(style),
                    style = TextStyle(
                        color = if (selected) colors.accent else colors.onSurface,
                        fontSize = (15 * style.relativeSize.coerceAtMost(1.6f)).sp,
                        fontWeight = FontWeight(style.weight)
                    )
                )
            }
        }
    }
}

@Composable
private fun ColorSheet(title: String, palette: List<String>, onColor: (String?) -> Unit) {
    val colors = SmartTextTokens.colors
    var picked by remember { mutableStateOf(SmartTextStyleMapper.parseColor(palette.first())) }
    SheetTitle(title)
    ColorSelector(initial = picked) { picked = it }
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(bottom = 20.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Button(
            onClick = { onColor(picked.toHexString()) },
            colors = ButtonDefaults.buttonColors(
                containerColor = colors.accent,
                contentColor = colors.onAccent
            )
        ) {
            Text("Apply")
        }
        TextButton(onClick = { onColor(null) }) {
            Text("Remove colour", color = colors.accent)
        }
    }
}

@Composable
private fun FontSizeSheet(current: Float?, onFontSize: (Float?) -> Unit) {
    val colors = SmartTextTokens.colors
    var size by remember { mutableFloatStateOf(current ?: 16f) }
    SheetTitle("Font size")
    Text("${size.toInt()} pt", style = TextStyle(color = colors.onSurfaceMuted, fontSize = 14.sp))
    Slider(
        value = size,
        onValueChange = { size = it },
        onValueChangeFinished = { onFontSize(size) },
        valueRange = 10f..48f,
        steps = 37,
        modifier = Modifier.padding(vertical = 8.dp)
    )
    TextButton(onClick = { onFontSize(null) }, modifier = Modifier.padding(bottom = 16.dp)) {
        Text("Reset to default", color = colors.accent)
    }
}

@Composable
private fun AlignSheet(current: TextAlign, onAlign: (TextAlign) -> Unit) {
    val colors = SmartTextTokens.colors
    SheetTitle("Alignment")
    Column(modifier = Modifier.padding(bottom = 20.dp)) {
        SmartTextCatalog.alignments.forEach { align ->
            val label = SmartTextCatalog.alignLabel(align)
            val selected = align == current
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 2.dp)
                    .background(
                        if (selected) colors.accentSoft else Color.Transparent,
                        RoundedCornerShape(8.dp)
                    )
                    .clickable { onAlign(align) }
                    .padding(horizontal = 12.dp, vertical = 12.dp)
            ) {
                Text(
                    text = label,
                    style = TextStyle(
                        color = if (selected) colors.accent else colors.onSurface,
                        fontSize = 15.sp
                    )
                )
            }
        }
    }
}

@Composable
private fun TableSheet(
    rowCount: Int,
    columnCount: Int,
    onInsertTable: (Int, Int) -> Unit,
    onTable: (TableCommand) -> Unit
) {
    val colors = SmartTextTokens.colors
    SheetTitle("Table")
    Row(
        modifier = Modifier.padding(bottom = 16.dp),
        horizontalArrangement = Arrangement.spacedBy(10.dp)
    ) {
        SmartTextCatalog.tablePresetRows.indices.forEach { index ->
            val rows = SmartTextCatalog.tablePresetRows[index]
            val columns = SmartTextCatalog.tablePresetColumns[index]
            Button(
                onClick = { onInsertTable(rows, columns) },
                colors = ButtonDefaults.buttonColors(
                    containerColor = colors.accentSoft,
                    contentColor = colors.accent
                )
            ) {
                Text("$rows × $columns")
            }
        }
    }
    Column(modifier = Modifier.padding(bottom = 20.dp)) {
        listOf(
            "Add row" to TableCommand.AddRow(rowCount),
            "Add column" to TableCommand.AddColumn(columnCount),
            "Delete last row" to TableCommand.DeleteRow(rowCount - 1),
            "Delete last column" to TableCommand.DeleteColumn(columnCount - 1),
            "Toggle header row" to TableCommand.ToggleHeaderRow,
            "Toggle header column" to TableCommand.ToggleHeaderColumn
        ).forEach { (label, command) ->
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .clickable { onTable(command) }
                    .padding(vertical = 12.dp)
            ) {
                Text(label, style = TextStyle(color = colors.onSurface, fontSize = 15.sp))
            }
        }
    }
}

@Composable
private fun LinkSheet(current: String?, onLink: (String?) -> Unit) {
    val colors = SmartTextTokens.colors
    var url by remember { mutableStateOf(current.orEmpty()) }
    SheetTitle(if (current == null) "Insert link" else "Edit link")
    OutlinedTextField(
        value = url,
        onValueChange = { url = it },
        placeholder = { Text("https://", color = colors.onSurfaceMuted) },
        singleLine = true,
        colors = TextFieldDefaults.colors(
            focusedContainerColor = Color.Transparent,
            unfocusedContainerColor = Color.Transparent,
            focusedIndicatorColor = colors.accent,
            unfocusedIndicatorColor = colors.divider,
            cursorColor = colors.accent,
            focusedTextColor = colors.onSurface,
            unfocusedTextColor = colors.onSurface
        ),
        modifier = Modifier.fillMaxWidth()
    )
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(top = 12.dp, bottom = 20.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Button(
            onClick = { onLink(url.takeIf { it.isNotBlank() }) },
            colors = ButtonDefaults.buttonColors(
                containerColor = colors.accent,
                contentColor = colors.onAccent
            )
        ) {
            Text("Apply")
        }
        if (current != null) {
            TextButton(onClick = { onLink(null) }) {
                Text("Remove link", color = colors.accent)
            }
        }
    }
}

@Composable
private fun FindReplaceSheet(
    query: String,
    replacement: String,
    matchCount: Int,
    currentMatch: Int,
    onSearchQuery: (String) -> Unit,
    onReplacement: (String) -> Unit,
    onFindNext: () -> Unit,
    onFindPrevious: () -> Unit,
    onReplaceCurrent: () -> Unit,
    onReplaceAll: () -> Unit
) {
    val colors = SmartTextTokens.colors
    SheetTitle("Find and replace")
    OutlinedTextField(
        value = query,
        onValueChange = onSearchQuery,
        placeholder = { Text("Find", color = colors.onSurfaceMuted) },
        singleLine = true,
        colors = sheetFieldColors(),
        modifier = Modifier.fillMaxWidth()
    )
    Text(
        text = if (matchCount == 0) "No matches" else "${currentMatch + 1} of $matchCount",
        style = TextStyle(color = colors.onSurfaceMuted, fontSize = 13.sp),
        modifier = Modifier.padding(vertical = 6.dp)
    )
    OutlinedTextField(
        value = replacement,
        onValueChange = onReplacement,
        placeholder = { Text("Replace with", color = colors.onSurfaceMuted) },
        singleLine = true,
        colors = sheetFieldColors(),
        modifier = Modifier.fillMaxWidth()
    )
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(top = 12.dp, bottom = 20.dp),
        horizontalArrangement = Arrangement.spacedBy(6.dp)
    ) {
        TextButton(onClick = onFindPrevious) { Text("Previous", color = colors.accent) }
        TextButton(onClick = onFindNext) { Text("Next", color = colors.accent) }
        TextButton(onClick = onReplaceCurrent) { Text("Replace", color = colors.accent) }
        TextButton(onClick = onReplaceAll) { Text("All", color = colors.accent) }
    }
}

@Composable
private fun sheetFieldColors() = TextFieldDefaults.colors(
    focusedContainerColor = Color.Transparent,
    unfocusedContainerColor = Color.Transparent,
    focusedIndicatorColor = SmartTextTokens.colors.accent,
    unfocusedIndicatorColor = SmartTextTokens.colors.divider,
    cursorColor = SmartTextTokens.colors.accent,
    focusedTextColor = SmartTextTokens.colors.onSurface,
    unfocusedTextColor = SmartTextTokens.colors.onSurface
)
